src/Control/Effect/Lift.hs

```diff
@@ -1,28 +1,46 @@
-{-# LANGUAGE FlexibleContexts, MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleContexts, FlexibleInstances, GeneralizedNewtypeDeriving, KindSignatures, MultiParamTypeClasses #-}
 module Control.Effect.Lift
 ( Lift(..)
 , sendM
 , runM
 , LiftC(..)
 ) where
 
+import Control.Applicative (Alternative(..))
 import Control.Effect.Carrier
 import Control.Effect.Sum
-import Control.Effect.Internal
-import Control.Effect.Lift.Internal
+import Control.Monad (MonadPlus(..))
+import Control.Monad.Fail
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import Data.Coerce
+
+newtype Lift sig (m :: * -> *) k = Lift { unLift :: sig k }
+  deriving (Functor)
+
+instance Functor sig => HFunctor (Lift sig) where
+  hmap _ = coerce
+  {-# INLINE hmap #-}
+
+instance Functor sig => Effect (Lift sig) where
+  handle state handler (Lift op) = Lift (fmap (handler . (<$ state)) op)
+
 
 -- | Extract a 'Lift'ed 'Monad'ic action from an effectful computation.
-runM :: Monad m => Eff (LiftC m) a -> m a
-runM = runLiftC . interpret
+runM :: LiftC m a -> m a
+runM = runLiftC
 
 -- | Given a @Lift n@ constraint in a signature carried by @m@, 'sendM'
 -- promotes arbitrary actions of type @n a@ to @m a@. It is spiritually
 -- similar to @lift@ from the @MonadTrans@ typeclass.
-sendM :: (Member (Lift n) sig, Carrier sig m, Functor n, Applicative m) => n a -> m a
+sendM :: (Member (Lift n) sig, Carrier sig m, Functor n) => n a -> m a
 sendM = send . Lift . fmap pure
 
 newtype LiftC m a = LiftC { runLiftC :: m a }
+  deriving (Alternative, Applicative, Functor, Monad, MonadFail, MonadIO, MonadPlus)
```

Comment:
Note that unlike `FailC`’s implementation of `MonadFail`, `LiftC`’s implementation of `MonadIO` represents an improvement to its semantics. Instead of providing `MonadIO` only when `Lift IO` is in the stack, we provide it any time `LiftC m` is in the stack for a type `m` which itself has a `MonadIO` instance.

This is kind of a neat instance of us using both `mtl`-style and effect-style semantics: in this case we get the improved type inference of `mtl` while still retaining the effectful interface (`Lift`, `sendM`) for other use cases.